### ISSUE TITLE
PyTorch Hub Inference error 

### ISSUE THREAD

**Issue by johanenforcer** [ultralytics/yolov5#2771]
## ❔Question
I'm currently using the PIL library for the purpose of image inference similarly to the way that the tutorial showed. That being said the error is at line  208, in common.py/forward, specifically _tuple index out of range_(if im.shape[0] < 5:  # image in CHW), and I am not quite sure what is going on. If I remove the im.close() after appending all requested images to a single list, it instead gives me the error that too many files have been opened, if I close them after appending them, it gives me the tuple index error. 

I have also tried using the LoadImages class but seems to give the same error.

Thanks in advance!

**Comment by glenn-jocher** [comment]
If you believe you have a reproducible issue, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. Thank you!

**Comment by johanenforcer** [comment]
> If you believe you have a reproducible issue, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. Thank you!

I'm not so sure about that, maybe this is a better explaination:

Im trying to use active learning in the context of object detection in a batch setting, meaning that there is an initial pool of unlabeled data (actually training data with corresponding labels) that I have to iteratively select 300 images from in many cycles of training. For each cycle these 300 images are selected and consequently incorporated into the actual "labelled" training set, used to train the model. In order to know which 300 images to select of the many training samples that exist, I need to run inference on (the entire training set available ("unlabeled pool") - already selected samples that exist in the "actual" training set) to evaluate which sampls are more useful to include in the "actual" training set. Perhaps wrongly, I chose to do this selection of 300 images in the dataset.py file, specifically in the LoadImagesAndLabels class as shown in the second image bellow. 

The image bellow shows the code, line 405 shows using PIL to append the "unlabeled samples" to feed the custom loaded pytorch hub model, according to how the tutorial does it. Currently only 1000 of these unlabeled samples are chosen for inference but even if I remove this constraint and allow the entire training set of approximately 5000 images to be chosen for inference, the problem doesn't change, which is that I get a certain error message, shown bellow.

Furthermore, something weird that I noticed, was that this error message happens if line 410, "im.close" is not commented out. If line 410, is commented out, everything works but I can only use very few images then since too many images have been opened. 

Error message:
![error3](https://user-images.githubusercontent.com/64362830/114561583-f9b9be00-9c6d-11eb-890d-c838f79c9628.png)

Image, showing the code:
![error5](https://user-images.githubusercontent.com/64362830/114564687-d7756f80-9c70-11eb-95ea-fe4b3e361016.png)

**Comment by glenn-jocher** [comment]
@johanenforcer I'm not sure I follow completely, but I would highly recommend not attempting inference with a batch-size of 1000 images.

**Comment by johanenforcer** [comment]
> @johanenforcer I'm not sure I follow completely, but I would highly recommend not attempting inference with a batch-size of 1000 images.

Hmm, it is probably pointless to ask this question but does it make a difference if it's the probability vectors that I'm after per bounding box (x[:, 5:]) because thats what I need to use.

**Comment by glenn-jocher** [comment]
Hmm, the class vectors are converted into detections during NMS, so you’d want to modify the NMS function to be able to extract them.

**Comment by johanenforcer** [comment]
> Hmm, the class vectors are converted into detections during NMS, so you’d want to modify the NMS function to be able to extract them.

Yea I did do that I think, but still receive the same error, It's weird because if I remove im.close() it seems to work for very few images like for example 300 instead of 1000 but with it, I get that error. Maybe it's what you said then? That the imgs list which is inputted for inference is not supposed to have a lot of images then? Is there a way that I can load thousands of images (maybe 3k at most, steadily decreasing) for inference to extract the class probability scores without using the PIL as the tutorial did? I tried to use the LoadImages class from the same dataset.py file but still received the same error message, weirdly enough.

**Comment by glenn-jocher** [comment]
@johanenforcer you can pass as many images as you want for inference.

If you want inference to actually succeed I would recommend limiting yourself to a common sense batch size << 1000 images, otherwise you will obviously run into CUDA OOM errors.
